README.md
smartcloud
===

Provides support for interacting with IBM SmartCloud API and CLI tools

Installation
===

from rubygems.org:

    gem install cft_smartcloud

locally:

    rake build
    gem install pkg/[name of generated gem]

Setup
===

Please set up SMARTCLOUD_USERNAME and SMARTCLOUD_PASSWORD in your .bash_profile

    export SMARTCLOUD_USERNAME=[your username]
    export SMARTCLOUD_PASSWORD=[your password]

You can now also supply the username and password on the command line using -u and -p
Use `smartcloud help` to get a list of all optoins.


Using the console
== 

    script/console
    >> @smartcloud.display_instances
    >> @smartcloud.display_volumes(:Location => 101)
    >> @smartcloud.display_instances(:Location => 82, :Name => "namematch")
    >> @smartcloud.poll_for_volume_state(12345, :unmounted)

  The console predefines the @smartcloud instance variable that is set up
  from the environment variables SMARTCLOUD_USERNAME and SMARTCLOUD_PASSWORD
  automatically (it's created at the bottom of smartcloud.rb)

Using the commandline 
== 

To see a list of methods:

    smartcloud help

Examples:

      smartcloud display_volumes
      smartcloud display_volumes Location=82 State=MOUNTED
      smartcloud display_instances
      smartcloud delete_instances 12345 12346 12347
      smartcloud display_images Name="Red Hat"
      smartcloud display_instances Name="Red Hat" Location=82

The 'display_*' methods are intended to generate pretty human readable 
displays, while the describe methods will return pretty-formatted hashes,
or singular values.

To save time when dealing with large responses, such as the describe_images
call, you can save a response in its native XML format:

      smartcloud display_images -S /tmp/images.xml

You can then replay the response, using filters on it

      smartcloud display_images Name='Red Hat' -R /tmp/images.xml



RestClient vs CurlHttpClient
===
Sometimes RestClient and friends have trouble communicating with certain API's such as 
IBM SmartCloud, returning 500 errors. We found in some cases the only thing that truly
works is pure curl (not even libcurl ruby wrappers). Therefore there is a provided 
simple CurlHttpClient library which emulates the RestClient interface, and wraps 
command line calls to curl.

The choice of client is determined inside of config.yml

Versioning
== 

This project uses the jeweler gem for packaging. See the tasks:

    rake version:bump:...
    rake build

Screencast
===

http://www.youtube.com/cohesiveft#p/u/0/-WdSHP2iwDM (somewhat outdated)

Copyright
== 

Copyright (c) 2011 CohesiveFT. See LICENSE for details.


CHANGELOG
2011-11-22 0.3.1:
 * Support for invocation of commands using key=value syntax
   Such as: smartcloud display_images name='Red Hat'

2011-11-21 0.3.0:
 * Support for smartcloud import/export api
 * Ability to use -S to save responses and -R to play them back 
	 for offline testing
 * Cleaner output display using ascii tables

2011-11-16 0.2.5: 
 * Better help display
 * Support for -u, -p, -U (username, password, api url)
 * Removed delete_unused_keys method (was not working correctly / not needed)

2011-09-21 0.2.1:
 * Added sleep/retry logic for intermittent 500 errors returned by cloud

2011-09-13 0.2.0:
 * Clone operation now done using API instead of CLI

2011-08-09 0.1.8:
 * Updated create_instance to return the Instance object rather 
   than the xml request, for consistency with describe_instance
 * Improved help system, especially for create_instance

2011-08-05 0.1.6:
 * support for deleting multiple volumes at once (delete_volumes)

2011-07-25 0.1.4:
 * Better display of launch times
 * delete_unused_keys function deletes all unused keys, prompting for each one
 * ability to find images by name display_images(:Name => "foo"), with fuzzy
   matching

2011-07-12 0.1.0:
 * Support for 'smartcloud' commandline tool wrapping the API


lib/mock_smartcloud.rb
#!/usr/bin/env ruby
#####################################################################################
# Copyright (c) 2011, Cohesive Flexible Technologies, Inc.
# This copyrighted material is the property of Cohesive Flexible Technologies and
# is subject to the license terms of the product it is contained within, whether 
# in text or compiled form.  It is licensed under the terms expressed in the 
# accompanying README and LICENSE files.
# 
# This program is AS IS and WITHOUT ANY WARRANTY; without even the implied warranty 
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#####################################################################################

class MockSmartCloud
  def create_instance(*args)
    describe_instance
  end

  def describe_image(*args)
    {"CreatedTime"=>"2011-02-25T01:22:57.000Z", "Owner"=>"SYSTEM", 
      "SupportedInstanceTypes"=>{"InstanceType"=>[
          {"Label"=>"Bronze 32 bit", "ID"=>"BRZ32.1/2048/60*175", 
            "Price"=>{"Rate"=>"0.21", "PricePerQuantity"=>"1", "UnitOfMeasure"=>"UHR  ", "CountryCode"=>"897", "CurrencyCode"=>"USD"}}, 
          {"Label"=>"Copper 32 bit", "ID"=>"COP32.1/2048/60", 
            "Price"=>{"Rate"=>"0.21", "PricePerQuantity"=>"1", "UnitOfMeasure"=>"UHR  ", "CountryCode"=>"897", "CurrencyCode"=>"USD"}}, 
          {"Label"=>"Silver 32 bit", "ID"=>"SLV32.2/4096/60*350", 
            "Price"=>{"Rate"=>"0.21", "PricePerQuantity"=>"1", "UnitOfMeasure"=>"UHR  ", "CountryCode"=>"897", "CurrencyCode"=>"USD"}}, 
        ]
      }, 
      "Name"=>"CohesiveFT VPN-Cubed Datacenter Connect V2.0.0 - BYOL (TOR)", "Location"=>"101", "ProductCodes"=>{"ProductCode"=>"caondc13bV97Kg5TLyGNodD9or7gA"}, "Documentation"=>"https://www-147.ibm.com/cloud/enterprise/ram.ws/RAMSecure/artifact/{974E595E-566E-87E3-A483-AC0FC468C4D9}/1.0/GettingStarted.html", "ID"=>"20016530", "Manifest"=>"https://www-147.ibm.com/cloud/enterprise/ram.ws/RAMSecure/artifact/{974E595E-566E-87E3-A483-AC0FC468C4D9}/1.0/parameters.xml", "Description"=>"CohesiveFT VPN-Cubed 2.0.0 UL - for TOR datacenter", "State"=>"AVAILABLE", "Visibility"=>"PUBLIC", "Platform"=>"Red Hat Enterprise Linux/5.4"}
  end

  def describe_instance(*args)
    {"Status"=>"NEW", "PrimaryIP"=>{"IP"=>{}, "Hostname"=>{}, "Type"=>"DYNAMIC"}, "ImageID"=>"20015391", "DiskSize"=>"60", "Owner"=>"smartes@cohesiveft.com", "RequestID"=>{"name"=>"sir_19_rhel5532-bit-1zGS", "content"=>"102271"}, "Name"=>"MockInstance", "ProductCodes"=>{}, "KeyName"=>"si-tempkey-rhel5532-bit-1zGS-eDY4T", "Location"=>"101", "LaunchTime"=>"2011-07-07T22:13:49.891Z", "Volumes"=>{}, "ID"=>"101971", "MiniEphemeral"=>"false", "Software"=>{"Application"=>{"Version"=>"5.5", "Name"=>"Red Hat Enterprise Linux", "Type"=>"OS"}}, "IP"=>{}, "InstanceType"=>"COP32.1/2048/60", "Hostname"=>{}}
  end

  def create_volume(*args)
    "12345"
  end

  def describe_volume(*args)
    {:ID => "12345", :State => "5", :Name => "foobar"}
  end

  def clone_volume(*args)
    "555555"
  end

  def restart_instance(*args)
    true
  end

  def delete_instance(*args)
    true
  end

  def poll_for_volume_state(*args)
    describe_volume(*args)
  end

  def poll_for_instance_state(*args)
    describe_instance(*args)
  end
end
